extra/migrate_settings.rb

```diff
@@ -1,108 +1,202 @@
+    end
+
+    def migration_dir
+      File.join(working_dir_path, migration_name)
+    end
+
+    def src_dir
+      File.join(migration_dir, "src")
+    end
+
+    def dst_dir
+      File.join(migration_dir, "dst")
+    end
+
+    def underscore(src)
+      src = src.gsub(/::/, '/')
+      src = src.gsub(/([A-Z]+)([A-Z][a-z])/,'\1_\2')
+      src = src.gsub(/([a-z\d])([A-Z])/,'\1_\2')
+      src = src.tr("-", "_")
+      src.downcase
     end
   end
 
-  # Rename the port to whichever is correct
-  if output[:settings].keys.include?(:port)
-    if output[:settings].keys.include?(:ssl_certificate)
-      output[:settings][:https_port] = output[:settings].delete(:port)
-    else
-      output[:settings][:http_port] = output[:settings].delete(:port)
+  class Migrator
+    attr_reader :working_dir_path, :migrations_dir_path, :config_src_dir, :result_dir_path
+
+    def initialize(working_dir_path, migrations_dir_path, config_src_dir, migrations)
+      @working_dir_path = working_dir_path
+      @migrations_dir_path = migrations_dir_path
+      @config_src_dir = config_src_dir
+      @migrations = migrations
+      @result_dir_path = File.join(working_dir_path, 'result')
+    end
+
+    def load_migrations(migrations_dir_path)
+      migration_files = Dir.glob(File.join(migrations_dir_path, "*.rb")).sort
+      ::Proxy::Migration.inject_migrations_instance(@migrations)
+      migration_files.each { |file| require file }
+    end
+
+    def migrate
+      raise "Migrations dir '#{migrations_dir_path}' doesn't exist" unless Dir.exist?(migrations_dir_path)
+      raise "Settings dir '#{config_src_dir}' doesn't exist" unless Dir.exist?(config_src_dir)
+
+      puts "using: \n config files dir: #{config_src_dir},\n working dir: #{working_dir_path},\n migrations dir: #{migrations_dir_path},\n migrations state path: #{@migrations.migration_state_file_path}"
+
+      FileUtils.mkdir_p(result_dir_path)
+      load_migrations(migrations_dir_path)
+
+      puts "Running migrations..."
+      migrations = @migrations.new_migrations
+      migrations.each do |migration|
+        m = migration.new(working_dir_path)
+        puts "#{m.migration_name}"
+
+        create_migration_dirs(m.src_dir, m.dst_dir)
+        if migration == migrations.first
+          copy_original_configuration(m.src_dir)
+        else
+          copy_previous_migration_results(@migrations.executed_migrations.last.dst_dir, m.src_dir)
+        end
+
+        m.migrate
+
+        if migration == migrations.last
+          copy_to_results_dir(m.dst_dir, result_dir_path)
+        end
+
+        @migrations.executed_migrations << m
+      end
+    ensure
+      @migrations.persist_migrations_state(result_dir_path)
+    end
+
+    def create_migration_dirs(src_dir, dst_dir)
+      FileUtils.mkdir_p(src_dir)
+      FileUtils.mkdir_p(File.join(dst_dir, "settings.d"))
+    end
+
+    def copy_original_configuration(src_dir)
+      FileUtils.cp_r(File.join(config_src_dir, '.'), src_dir)
```

Comment:
This doesn't quite work with our packaging, as the settings.d dir can be changed by `:settings_directory: /etc/foreman-proxy/settings.d` in the settings.yml file.  You might need to open the settings.yml and check that setting to copy it too.